File: leetcode/max_points_on_line.py
# Given n points on a 2D plane, find the maximum number of points that lie on 
# the same straight line.
# Time Complexity: O(n^2) in average
# Space Complexity: O(n)

# Definition for a point.
class Point:
    def __init__(self, a=0, b=0):
        self.x = a
        self.y = b

from fractions import Fraction, gcd

class Solution:
    # @param {Point[]} points
    # @return {integer}
    def maxPoints(self, points):
        result = 0
        for p1 in points:
            dict, count = {}, 0
            for p2 in points:
                dx, dy = p2.x - p1.x, p2.y - p1.y
                if dx == 0 == dy: count += 1
                else: dy, dx = dy / gcd(dx, dy), dx / gcd(dx, dy)
                m = (dy, dx) if dx else None
                dict[m] = dict.get(m, 0) + 1
            dict[None] -= count
            result = max(result, count + max(dict.viewvalues()))
        return result

if __name__ == "__main__":
    solution = Solution()
    print solution.maxPoints([Point(0, 0), Point(0, 0)])

File: leetcode/max_points_on_line.js
/* Given n points on a 2D plane, find the maximum number of points that lie on 
 * the same straight line.
 * Time Complexity: O(n^2) in average
 * Space Complexity: O(n)
 */

/**
 * Definition for a point.
 * function Point(x, y) {
 *     this.x = x;
 *     this.y = y;
 * }
 */
/**
 * @param {Point[]} points
 * @return {number}
 */
var maxPoints = function(points) {
    function gcd(a, b) {
        while ((a %= b) && (b %= a))
            ;
        return a + b
    }
    for (var result = 0, i = 0; i < points.length; ++i) {
        dict = {}
        for (var count = 0, j = 0; j < points.length; ++j) {
            var dx = points[j].x - points[i].x
            var dy = points[j].y - points[i].y
            if (dx == 0 && dy == 0)
                ++count
            else {
                var d = gcd(dx, dy)
                dx /= d
                dy /= d
            }
            var m = dx != 0 ? [dy, dx] : null
            if (m in dict)
                dict[m] = dict[m] + 1
            else
                dict[m] = 1
        }
        dict[null] -= count
        for(var key in dict) {
            if(count + dict[key] > result)
                result = count + dict[key]
        }
    }
    return result
};
